### PR TITLE
add gas fee payer by adding a new payload type

### PR DESCRIPTION
**Pull request by gerben-stavenga**
### Description

Add gas payer by adding a entry to RawMessageWithData however this struct is not being transmitted. So we need to add a new entry in the payload struct to transmit the intend of gas payer.

### CODE REVIEW

File: types/src/transaction/mod.rs [hunk 1]
```diff
@@ -366,6 +368,10 @@ pub enum RawTransactionWithData {
         raw_txn: RawTransaction,
         secondary_signer_addresses: Vec<AccountAddress>,
     },
+    MultiAgentWithFeePayer {
+        raw_txn: RawTransaction,
+        secondary_signer_addresses: Vec<AccountAddress>,
```
Comment (davidiw) [inline types/src/transaction/mod.rs:373]: should this not include an additional signer called `fee_payer: Option<AccountAddress>` or maybe no option?

Comment (gerben-stavenga) [reply types/src/transaction/mod.rs:376]: It's still not clear how to get fee_payer into the system. But we can always treat it as just another agent

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 2]
```diff
@@ -513,9 +516,12 @@ impl AptosVMImpl {
         txn_data: &TransactionMetadata,
         transaction_validation: &TransactionValidation,
     ) -> VMResult<()> {
-        let txn_sequence_number = txn_data.sequence_number();
+        let mut txn_sequence_number = txn_data.sequence_number();
         let txn_gas_price = txn_data.gas_unit_price();
         let txn_max_gas_units = txn_data.max_gas_amount();
+        if txn_data.with_gas_payer {
+            txn_sequence_number |= GAS_PAYER_FLAG_BIT;
+        }
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/aptos_vm_impl.rs:524]: should this still be there?

File: aptos-move/aptos-vm/src/transaction_metadata.rs [hunk 3]
```diff
@@ -17,6 +17,7 @@ pub struct TransactionMetadata {
     pub secondary_signers: Vec<AccountAddress>,
     pub secondary_authentication_keys: Vec<Vec<u8>>,
     pub sequence_number: u64,
+    pub with_gas_payer: bool,
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/transaction_metadata.rs:20]: I don't understand why this isn't an explicit field.

File: types/src/transaction/mod.rs (hunk 1 above)
Comment (davidiw) [reply types/src/transaction/mod.rs:382]: You mean `TransactionAuthenticator`? Yeah, I think we need a new one of these, this is still hacky.

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 2 above)
Comment (gerben-stavenga) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:524]: It's used to communicate with aptos-framework

File: aptos-move/aptos-vm/src/transaction_metadata.rs (hunk 3 above)
Comment (gerben-stavenga) [reply aptos-move/aptos-vm/src/transaction_metadata.rs:20]: I think it's good to work towards an orthogonal primitives, instead of adding special cases for each new hot feature we think of. We have a authenticator for multiple signatures, either we reuse it or we implement a 3rd special cased with potential bugs.

File: aptos-move/aptos-vm/src/transaction_metadata.rs [hunk 4]
```diff
@@ -41,8 +41,11 @@ impl TransactionMetadata {
                 .map(|account_auth| account_auth.authentication_key().to_vec())
                 .collect(),
             sequence_number: txn.sequence_number(),
```
Comment (gerben-stavenga) [inline aptos-move/aptos-vm/src/transaction_metadata.rs:43]: Because we don't do the sequence number bit trick we need feature flags now in this code to prevent forking

**Review by gerben-stavenga — COMMENTED** [review]
lgtm

File: aptos-move/aptos-vm/src/transaction_metadata.rs (hunk 4 above)
Comment (gerben-stavenga) [reply aptos-move/aptos-vm/src/transaction_metadata.rs:43]: okay if both path are surely not going on-chain then it's fine

File: aptos-move/aptos-vm/src/errors.rs [hunk 5]
```diff
@@ -156,6 +156,7 @@ pub fn convert_epilogue_error(
     log_context: &AdapterLogSchema,
 ) -> Result<(), VMStatus> {
     let status = error.into_vm_status();
+    println!("{:?}", status);
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/errors.rs:159]: nit

File: ecosystem/indexer-grpc/indexer-grpc-parser/src/models/default_models/signatures.rs [hunk 6]
```diff
@@ -9,12 +9,13 @@ use anyhow::{Context, Result};
...
     account_signature::Signature as AccountSignatureEnum, signature::Signature as SignatureEnum,
     AccountSignature as ProtoAccountSignature, Ed25519Signature as Ed25519SignaturePB,
-    MultiAgentSignature as ProtoMultiAgentSignature,
+    FeePayerSignature as ProtoFeePayerSignature, MultiAgentSignature as ProtoMultiAgentSignature,
     MultiEd25519Signature as ProtoMultiEd25519Signature, Signature as TransactionSignaturePB,
 };
 use field_count::FieldCount;
 use serde::{Deserialize, Serialize};
 
+// TODO: fee payer is just an extended multiagent with the last entry being the fee payer
```
Comment (davidiw) [inline ecosystem/indexer-grpc/indexer-grpc-parser/src/models/default_models/signatures.rs:18]: nit

File: sdk/src/types.rs [hunk 7]
```diff
@@ -105,11 +105,11 @@ impl LocalAccount {
         secondary_signers: Vec<&Self>,
         builder: TransactionBuilder,
     ) -> SignedTransaction {
-        let secondary_signer_addresses = secondary_signers
+        let secondary_signer_addresses: Vec<AccountAddress> = secondary_signers
             .iter()
             .map(|signer| signer.address())
             .collect();
-        let secondary_signer_privkeys = secondary_signers
+        let secondary_signer_privkeys: Vec<&Ed25519PrivateKey> = secondary_signers
```
Comment (davidiw) [inline sdk/src/types.rs:112]: probably can drop these changes

File: types/src/transaction/authenticator.rs [hunk 8]
```diff
@@ -46,6 +46,14 @@ pub enum TransactionAuthenticator {
         public_key: Ed25519PublicKey,
         signature: Ed25519Signature,
     },
+    /// Optional Multi-agent transaction with a fee payer.
+    FeePayer {
+        sender: AccountAuthenticator,
```
Comment (davidiw) [inline types/src/transaction/authenticator.rs:51]: yup, ordering here matters, I broke bcs...

**Review by davidiw — APPROVED** [review]
good with the few tweaks and of course cicd passing

File: api/src/transactions.rs [hunk 9]
```diff
@@ -852,7 +852,10 @@ impl TransactionsApi {
                         })?;
                 // Verify the signed transaction
                 match signed_transaction.payload() {
-                    TransactionPayload::EntryFunction(entry_function) => {
+                    TransactionPayload::MultiAgentWithFeePayer(
+                        MultisigTransactionPayload::EntryFunction(entry_function),
```
Comment (movekevin) [inline api/src/transactions.rs:856]: This makes it impossible to support scripts with fee payer in the future

File: aptos-move/e2e-move-tests/src/tests/scripts.rs [hunk 10]
```diff
@@ -84,7 +84,7 @@ fn read_coin(h: &MoveHarness, account: &AccountAddress) -> u64 {
 }
 
 #[test]
-fn test_two_to_two_transfer_gas_payer() {
+fn test_two_to_two_transfer_fee_payer() {
```
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/tests/scripts.rs:87]: Let's actually move these tests out of scripts.rs into their own file. It has been confusing to find tests as we put too many of them in here even if they're not that strongly related

Comment (gerben-stavenga) [reply aptos-move/e2e-move-tests/src/tests/scripts.rs:87]: Can we do that in follow-up? I added them here as they are the only one that was using multi agent.

File: api/types/src/transaction.rs [hunk 11]
```diff
@@ -1228,6 +1235,89 @@ impl
...
+}
+
+impl VerifyInput for FeePayerSignature {
+    fn verify(&self) -> anyhow::Result<()> {
+        self.sender.verify()?;
+
+        for signer in self.secondary_signers.iter() {
+            signer.verify()?;
+        }
+        self.fee_payer_signer.verify()?;
```
Comment (movekevin) [inline api/types/src/transaction.rs:1261]: Do we need to verify that this signature matches the fee payer address here? Should at least be done in the execution flow somewhere

File: aptos-move/aptos-vm/src/transaction_metadata.rs [hunk 12]
```diff
@@ -100,7 +118,7 @@ impl TransactionMetadata {
     }
 
     pub fn is_multi_agent(&self) -> bool {
-        !self.secondary_signers.is_empty()
+        !(self.secondary_signers.is_empty() && self.fee_payer.is_none())
```
Comment (movekevin) [inline aptos-move/aptos-vm/src/transaction_metadata.rs:121]: This seems weird. Why is this necessary when fee payer is a separate tx?

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs [hunk 13]
```diff
@@ -163,6 +163,23 @@ pub fn convert_transaction_payload(
                 convert_multisig_payload(mp),
             )),
         },
+        TransactionPayload::FeePayerPayload(MultisigTransactionPayload::EntryFunctionPayload(sfp)) => {
```
Comment (movekevin) [inline ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:166]: I guess we won't know if this works until it's actually run :| cc @bowenyang007

File: api/types/src/transaction.rs (hunk 11 above)
Comment (gerben-stavenga) [reply api/types/src/transaction.rs:1261]: It's done in the prologue

File: aptos-move/aptos-vm/src/transaction_metadata.rs (hunk 12 above)
Comment (gerben-stavenga) [reply aptos-move/aptos-vm/src/transaction_metadata.rs:120]: @davidiw what do you think is best here

File: types/src/transaction/mod.rs [hunk 14]
```diff
@@ -283,6 +283,63 @@ impl RawTransaction {
...
+    /// The order and length of the secondary keys provided here have to match the order and
+    /// length of the `secondary_signers`.
+    pub fn sign_fee_payer(
+        self,
+        sender_private_key: &Ed25519PrivateKey,
+        secondary_signers: Vec<AccountAddress>,
+        secondary_private_keys: Vec<&Ed25519PrivateKey>,
+        fee_payer_address: AccountAddress,
+        fee_payer_private_key: &Ed25519PrivateKey,
+    ) -> Result<SignatureCheckedTransaction> {
```
Comment (gregnazario) [inline types/src/transaction/mod.rs:299]: You know, eventually we'll probably want to make these signatures stacked together later, rather than have all the private keys in one place (since they should never all be in the same place)

File: types/src/transaction/authenticator.rs [hunk 15]
```diff
@@ -191,6 +273,33 @@ impl fmt::Display for TransactionAuthenticator {
...
+            Self::FeePayer {
+                sender,
+                secondary_signer_addresses,
+                secondary_signers,
+                fee_payer_address,
+                fee_payer_signer,
+            } => {
+                let mut sec_addrs: String = "".to_string();
+                for sec_addr in secondary_signer_addresses {
+                    sec_addrs = format!("{}\n\t\t\t{:#?},", sec_addrs, sec_addr);
```
Comment (gregnazario) [inline types/src/transaction/authenticator.rs:285]: probably 2 tabs was enough, but luckily 8 char tabs aren't common anymore

File: aptos-move/framework/aptos-framework/sources/transaction_validation.spec.move [hunk 16]
```diff
@@ -111,42 +135,48 @@ spec aptos_framework::transaction_validation {
...
-        ensures forall i in 0..num_secondary_signers:
-            account::exists_at(secondary_signer_addresses[i])
-                && secondary_signer_public_key_hashes[i] ==
-                    account::get_authentication_key(secondary_signer_addresses[i]);
+        aborts_if !account::exists_at(gas_payer);
+        aborts_if !(fee_payer_public_key_hash == account::get_authentication_key(gas_payer));
     }
 
-    /// Abort according to the conditions.
+        /// Abort according to the conditions.
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.spec.move:179]: Nit: Weird tab here

File: aptos-move/e2e-move-tests/src/tests/scripts.rs [hunk 17]
```diff
@@ -82,341 +80,3 @@ fn read_coin(h: &MoveHarness, account: &AccountAddress) -> u64 {
         .unwrap()
         .coin()
 }
-
-#[test]
-fn test_two_to_two_transfer_gas_payer() {
-    let mut h = MoveHarness::new_with_features(vec![FeatureFlag::GAS_PAYER_ENABLED], vec![]);
-
-    let alice = h.new_account_at(AccountAddress::from_hex_literal("0xa11ce").unwrap());
```
Comment (gregnazario) [inline aptos-move/e2e-move-tests/src/tests/scripts.rs:90]: I guess we moved these?

File: aptos-move/aptos-vm/src/transaction_metadata.rs (hunk 12 above)
Comment (gregnazario) [reply aptos-move/aptos-vm/src/transaction_metadata.rs:120]: This isn't actually checking if it's `multi-agent`, but if it's `multi-signer` or `multi-agent`.  Can you verify where it's using this function?

File: aptos-move/e2e-move-tests/src/tests/fee_payer.rs [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        .script(script)
+        .sequence_number(h.sequence_number(alice.address()))
+        .max_gas_amount(1_000_000)
+        .gas_unit_price(1)
+        .sign_fee_payer();
+
+    let output = h.executor.execute_transaction(transaction);
+    assert!(transaction_status_eq(
+        output.status(),
+        &TransactionStatus::Discard(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION)
```
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/tests/fee_payer.rs:142]: Hmm this is getting UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION because prologue fails. This doesn't really look right to me at all as UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION is an error reserved for internal system (usually unexpected errors). This should return a validation error instead, similar to what I did with multisig prologue errors such as MULTISIG_TRANSACTION_NOT_FOUND

File: aptos-move/e2e-move-tests/src/tests/fee_payer.rs [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+    let transaction = TransactionBuilder::new(alice.clone())
+        .fee_payer(bob.clone())
+        .entry_function(entry)
+        .sequence_number(h.sequence_number(alice.address()))
+        .max_gas_amount(1_000_000)
+        .gas_unit_price(1)
+        .sign_fee_payer();
+
+    let output = h.executor.execute_transaction(transaction);
+    // The last signer became the gas payer and thus, the execution errors with a mismatch
```
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/tests/fee_payer.rs:174]: This comment is outdated.

File: aptos-move/e2e-move-tests/src/tests/fee_payer.rs [hunk 20]
```diff
@@ -0,0 +1,249 @@
+// Copyright © Aptos Foundation
```
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/tests/fee_payer.rs:1]: I think we need more tests for more cases such as:
1. Invalid gas payer signature
2. All valid signatures from senders and gas payers but signed over RawTransaction, now RawTransactionWithData
3. Gas payer account is a single signer (aka I pay for my own gas). I think most of the tests here only test when the payer is separate

File: types/src/transaction/authenticator.rs [hunk 21]
```diff
@@ -108,6 +133,25 @@ impl TransactionAuthenticator {
...
                 signature,
             } => signature.verify(raw_txn, public_key),
+            Self::FeePayer {
+                sender,
+                secondary_signer_addresses,
+                secondary_signers,
+                fee_payer_address,
+                fee_payer_signer,
+            } => {
+                let message = RawTransactionWithData::new_fee_payer(
```
Comment (movekevin) [inline types/src/transaction/authenticator.rs:143]: Can you leave a comment here that this is similar to multiagent transactions and is necessary so that the fee payer portion cannot be "stripped" out of a transaction that has been signed by senders?

File: aptos-move/aptos-vm/src/transaction_metadata.rs (hunk 12 above)
Comment (movekevin) [reply aptos-move/aptos-vm/src/transaction_metadata.rs:120]: I assume this is needed to know which prologue Move function to call since multiagent txs can now have separate gas payer as well, which means different prologue functions need to call for "old" multiagent txs (no gas payer) vs new. Is this necessary though as the vm logic can just look at the presence of gas payer + whether something is a multiagent tx to decide which function to call?

File: api/types/src/transaction.rs (hunk 11 above)
Comment (movekevin) [reply api/types/src/transaction.rs:1257]: Found this in Move in fee_payer_script_prologue. I wonder if we should have a separate error here because it'd be hard to know which signature mismatches with the address - senders' or payer's. Users would see the same error for both cases and be confused

**Review by movekevin — COMMENTED** [review]
Can you add a success test in api as well just to make sure the API portion works? E2E move test only test VM and execution. This now needs the API to work correctly as well

File: aptos-move/e2e-move-tests/src/tests/scripts.rs (hunk 17 above)
Comment (movekevin) [reply aptos-move/e2e-move-tests/src/tests/scripts.rs:90]: Yes they got moved to fee_payer. These shouldn't have been here in the first place. scripts.rs seems to getting a lot of random tests that don't really belong

File: aptos-move/framework/aptos-framework/sources/transaction_validation.move [hunk 22]
```diff
@@ -44,7 +43,8 @@ module aptos_framework::transaction_validation {
     const PROLOGUE_EBAD_CHAIN_ID: u64 = 1007;
     const PROLOGUE_ESEQUENCE_NUMBER_TOO_BIG: u64 = 1008;
     const PROLOGUE_ESECONDARY_KEYS_ADDRESSES_COUNT_MISMATCH: u64 = 1009;
-    const PROLOGUE_EGAS_PAYER_ACCOUNT_MISSING: u64 = 1010;
+    const PROLOGUE_EFEE_PAYER_NOT_ENABLED: u64 = 1010;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/transaction_validation.move:46]: As mentioned in the test earlier, this needs to be properly mapped to a non-internal error similar to what I did for the new multisig account

File: types/src/transaction/mod.rs (hunk 14 above)
Comment (movekevin) [reply types/src/transaction/mod.rs:299]: What do you mean by "stacked together"?

**Review by movekevin — APPROVED** [review]

File: aptos-move/e2e-move-tests/src/tests/fee_payer.rs (hunk 20 above)
Comment (gerben-stavenga) [reply aptos-move/e2e-move-tests/src/tests/fee_payer.rs:1]: I can't find a way to test for invalid signatures. As everything is private or crate private and the only way to construct signatures are valid signatures

Comment (zekun000) [reply aptos-move/e2e-move-tests/src/tests/fee_payer.rs:1]: https://github.com/aptos-labs/aptos-core/blob/main/crates/aptos-crypto/src/ed25519/ed25519_sigs.rs#L43
